### ISSUE TITLE
panic: runtime error: invalid memory address or nil pointer dereference (part.go:413)

### ISSUE THREAD

**Issue by nwoodmsft** [vbauerster/getparty#32]
This seems rare, but I noticed this nil pointer deref in one of my downloads. This was on getparty 1.22.1 release, but may still be present in 1.22.2 release.

Unfortunately, I do not have the exact steps to reproduce the issue, but it was a scenario where a multi-part download had failed all retries for 1 parrt (so a json session file was written) and then the download for the failed part was resumed from the json session.

Looks like a nil pointer deref inside the Part.writeTo function (part.go:413) when it tries to call the following in the defer block:

```
p.logger.Printf("Removing: %q", src.Name())
```

I think the src is a pointer, so there seems to be a case where src can be nil when this defer block executes the Printf call.

```
panic: runtime error: invalid memory address or nil pointer dereference
	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x24 pc=0x684d34]

goroutine 296 [running]:
testing.tRunner.func1.2({0x96c800, 0xd7cff0})
	/opt/hostedtoolcache/go/1.22.4/x64/src/testing/testing.go:1631 +0x24a
testing.tRunner.func1()
	/opt/hostedtoolcache/go/1.22.4/x64/src/testing/testing.go:1634 +0x377
panic({0x96c800?, 0xd7cff0?})
	/opt/hostedtoolcache/go/1.22.4/x64/src/runtime/panic.go:770 +0x132
log.(*Logger).output(0xc00021ed20?, 0x2e?, 0x0?, 0xc0005af978?)
	/opt/hostedtoolcache/go/1.22.4/x64/src/log/log.go:203 +0x34
log.(*Logger).Printf(...)
	/opt/hostedtoolcache/go/1.22.4/x64/src/log/log.go:268
github.com/vbauerster/getparty.Part.writeTo.func1()
	/home/vsts/go/pkg/mod/github.com/vbauerster/getparty@v1.22.1-0.20240805105734-269e8565947e/part.go:413 +0x18e
panic({0x96c800?, 0xd7cff0?})
	/opt/hostedtoolcache/go/1.22.4/x64/src/runtime/panic.go:770 +0x132
log.(*Logger).output(0xc00040b800?, 0x2b?, 0xac7da0?, 0xc0001383e8?)
	/opt/hostedtoolcache/go/1.22.4/x64/src/log/log.go:203 +0x34
log.(*Logger).Printf(...)
	/opt/hostedtoolcache/go/1.22.4/x64/src/log/log.go:268
github.com/vbauerster/getparty.Part.writeTo({0x0, 0x280525d, 0x280525e, 0x150f230a, {0x0, 0x0}, 0x0, 0x0, 0x0, 0x0, ...}, ...)
	/home/vsts/go/pkg/mod/github.com/vbauerster/getparty@v1.22.1-0.20240805105734-269e8565947e/part.go:425 +0x347
github.com/vbauerster/getparty.Session.concatenateParts({0x1, {0xc0001cc0f0, 0xed}, {0xc0001a6780, 0xed}, {0xc00040b800, 0x2b}, {0x0, 0x0}, {0xc00049ec48, ...}, ...}, ...)
	/home/vsts/go/pkg/mod/github.com/vbauerster/getparty@v1.22.1-0.20240805105734-269e8565947e/session.go:190 +0x1098
github.com/vbauerster/getparty.(*Cmd).Run(0xc00017e4d0, {0xc000188100, 0x6, 0x8}, {0x0, 0x0}, {0x0, 0x0})
	/home/vsts/go/pkg/mod/github.com/vbauerster/getparty@v1.22.1-0.20240805105734-269e8565947e/getparty.go:352 +0x1ccc

```

**Comment by vbauerster** [comment]
Thanks for reporting. I've fixed this in e25ec2c62526bcb48a31f715721f4eb00d412995, you're welcome to test of course if you have some spare time. To reproduce you need to start session `getparty -s session.json` with at least one completed bar.

**Comment by nwoodmsft** [comment]
Ah, it was the logger :) Makes sense - thank you for fixing this! I see that this commit made it into the v1.22.2 release, so I will move up to that version.
